### This project aims to predict the size category of a company based on the alumni’s professional data, such as their industry, company specialties, and LinkedIn connection count.



In [ ]:
# import pandas as pd
from sqlalchemy import create_engine

# PostgreSQL Connection
engine = create_engine("postgresql+psycopg2://postgres:postgres123@localhost:5432/DW")

# Loading the data
query = """
SELECT f."factKey", f."dateFK", f."AluminiFK", f."CompanyFK", f."num_of_connections",
       a."first_name", a."last_name", a."Industry", a."Current_Position",
       c."company_size", c."company_size_category", c."company_specialties"
FROM "Fact_Employability" f
JOIN "Dim_Alumini" a ON f."AluminiFK" = a."AluminiKey"
JOIN "Dim_Company" c ON f."CompanyFK" = c."CompanyKey"
"""
conn = engine.raw_connection()
df = pd.read_sql(query, conn)
conn.close()

# Data Cleaning
df['num_of_connections'] = pd.to_numeric(df['num_of_connections'], errors='coerce')
df = df.dropna(subset=['num_of_connections'])
df = df.drop(columns=["factKey", "first_name", "last_name", "Current_Position"])


In [2]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.pipeline import Pipeline
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import OneHotEncoder, StandardScaler, LabelEncoder
from sklearn.metrics import classification_report, accuracy_score
from sklearn.ensemble import RandomForestClassifier, GradientBoostingClassifier
from xgboost import XGBClassifier

# Data
X = df[['num_of_connections', 'Industry', 'company_specialties', 'company_size']]
y = df['company_size_category']

# Data Cleaning
X['num_of_connections'] = pd.to_numeric(X['num_of_connections'], errors='coerce')
X = X.dropna()
y = y.loc[X.index]

# Columns
categorical = ['Industry', 'company_specialties', 'company_size']
numerical = ['num_of_connections']

# Split
X_train, X_test, y_train_str, y_test_str = train_test_split(X, y, stratify=y, random_state=42)

# ✅ Creating an encoded version of y for XGBoost
label_encoder = LabelEncoder()
y_train_enc = label_encoder.fit_transform(y_train_str)
y_test_enc = label_encoder.transform(y_test_str)

# Results
results = []

# Model and preprocessor definitions
model_defs = {
    "RandomForest": (
        RandomForestClassifier(class_weight='balanced', random_state=42),
        ColumnTransformer([
            ("cat", OneHotEncoder(handle_unknown='ignore', sparse=True), categorical),
            ("num", StandardScaler(), numerical)
        ]),
        y_train_str,
        y_test_str
    ),
    "GradientBoosting": (
        GradientBoostingClassifier(random_state=42),
        ColumnTransformer([
            ("cat", OneHotEncoder(handle_unknown='ignore', sparse=True), categorical),
            ("num", StandardScaler(), numerical)
        ]),
        y_train_str,
        y_test_str
    ),
    "XGBoost": (
        XGBClassifier(use_label_encoder=False, eval_metric='mlogloss', random_state=42),
        ColumnTransformer([
            ("cat", OneHotEncoder(handle_unknown='ignore', sparse_output=False), categorical),
            ("num", StandardScaler(), numerical)
        ]),
        y_train_enc,
        y_test_enc
    )
}

# 📊 Training Loop
for name, (model, preprocessor, y_train_model, y_test_model) in model_defs.items():
    pipe = Pipeline([
        ("prep", preprocessor),
        ("clf", model)
    ])
    pipe.fit(X_train, y_train_model)
    y_pred = pipe.predict(X_test)
    
    acc = accuracy_score(y_test_model, y_pred)
    print(f"\n📊 {name} - Accuracy: {acc:.2%}")
    
    if name == "XGBoost":
        print(classification_report(y_test_model, y_pred, target_names=label_encoder.classes_))
    else:
        print(classification_report(y_test_model, y_pred))
    
    results.append((name, acc))

# Summary
df_results = pd.DataFrame(results, columns=["Model", "Accuracy"]).sort_values(by="Accuracy", ascending=False)
print("\n🎯 Model Summary:")
print(df_results.to_string(index=False))


C:\Users\ASUS\AppData\Local\Temp/ipykernel_14208/745332196.py:15: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  X['num_of_connections'] = pd.to_numeric(X['num_of_connections'], errors='coerce')
C:\Users\ASUS\anaconda3\lib\site-packages\sklearn\preprocessing\_encoders.py:868: FutureWarning: `sparse` was renamed to `sparse_output` in version 1.2 and will be removed in 1.4. `sparse_output` is ignored unless you leave `sparse` to its default value.
  warnings.warn(



📊 RandomForest - Accuracy: 77.69%
              precision    recall  f1-score   support

        1-10       0.00      0.00      0.00         3
      10000+       0.57      0.99      0.73        70
   1001-2000       1.00      0.89      0.94        64
   1001-5000       1.00      0.78      0.88        37
       11-50       1.00      0.18      0.31        11
     201-500       1.00      0.30      0.46        10
  5001-10000       1.00      0.20      0.33         5
    501-1000       1.00      0.60      0.75         5
      51-200       0.89      0.65      0.75        37

    accuracy                           0.78       242
   macro avg       0.83      0.51      0.57       242
weighted avg       0.85      0.78      0.76       242



C:\Users\ASUS\anaconda3\lib\site-packages\sklearn\metrics\_classification.py:1344: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
C:\Users\ASUS\anaconda3\lib\site-packages\sklearn\metrics\_classification.py:1344: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
C:\Users\ASUS\anaconda3\lib\site-packages\sklearn\metrics\_classification.py:1344: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
C:\Users\ASUS\anaconda3\lib\site-packages\sklearn\preproce


📊 GradientBoosting - Accuracy: 82.23%
              precision    recall  f1-score   support

        1-10       0.00      0.00      0.00         3
      10000+       0.62      1.00      0.77        70
   1001-2000       1.00      0.91      0.95        64
   1001-5000       1.00      0.78      0.88        37
       11-50       1.00      0.45      0.62        11
     201-500       1.00      0.50      0.67        10
  5001-10000       1.00      0.20      0.33         5
    501-1000       1.00      0.60      0.75         5
      51-200       1.00      0.76      0.86        37

    accuracy                           0.82       242
   macro avg       0.85      0.58      0.65       242
weighted avg       0.88      0.82      0.82       242



C:\Users\ASUS\anaconda3\lib\site-packages\sklearn\metrics\_classification.py:1344: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
C:\Users\ASUS\anaconda3\lib\site-packages\sklearn\metrics\_classification.py:1344: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
C:\Users\ASUS\anaconda3\lib\site-packages\sklearn\metrics\_classification.py:1344: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))



📊 XGBoost - Accuracy: 72.31%
              precision    recall  f1-score   support

        1-10       0.00      0.00      0.00         3
      10000+       0.70      0.89      0.78        70
   1001-2000       0.93      0.84      0.89        64
   1001-5000       0.88      0.78      0.83        37
       11-50       0.22      0.18      0.20        11
     201-500       0.12      0.10      0.11        10
  5001-10000       0.00      0.00      0.00         5
    501-1000       0.12      0.20      0.15         5
      51-200       0.72      0.70      0.71        37

    accuracy                           0.72       242
   macro avg       0.41      0.41      0.41       242
weighted avg       0.71      0.72      0.71       242


🎯 Model Summary:
           Model  Accuracy
GradientBoosting  0.822314
    RandomForest  0.776860
         XGBoost  0.723140


=> GradientBoosting has the highest accuracy at 82.23%, indicating it performed the best in classifying the data correctly.

RandomForest follows with an accuracy of 77.69%, showing a good performance, though slightly lower than GradientBoosting.

XGBoost has the lowest accuracy at 72.31%,